qos/evm/endpoint.go

```diff
@@ -2,130 +2,53 @@ package evm
-	// These endpoints are excluded from selection until service restart.
 	hasReturnedEmptyResponse bool
-
-	// chainIDResponse stores the result of processing the endpoint's response to an `eth_chainId` request.
-	// It is nil if there has NOT been an observation of the endpoint's response to an `eth_chainId` request.
-	chainIDResponse *string
-
-	// parsedBlockNumberResponse stores the result of processing the endpoint's response to an `eth_blockNumber` request.
-	// It is nil if there has NOT been an observation of the endpoint's response to an `eth_blockNumber` request.
-	parsedBlockNumberResponse *uint64
-
-	// observedArchivalBalance stores the result of processing the endpoint's response
-	// to an `eth_getBalance` request for a specific contract address at a specific block number.
-	observedArchivalBalance string
-}
-
-func (e *endpoint) validateEmptyResponse() error {
-	if e.hasReturnedEmptyResponse {
-		return errHasReturnedEmptyResponse
-	}
-	return nil
-}
-
-func (e *endpoint) validateChainID(chainID string) error {
-	if e.chainIDResponse == nil {
-		return errNoChainIDObs
-	}
-	if *e.chainIDResponse != chainID {
-		return fmt.Errorf("%s. expected: %s, got: %s", errInvalidChainIDObs, chainID, *e.chainIDResponse)
-	}
-	return nil
+	checkBlockNumber         endpointCheckBlockNumber
+	checkChainID             endpointCheckChainID
+	checkArchival            endpointCheckArchival
 }
 
-// validateBlockNumber checks if the perceived block number is valid.
-// The perceived block number:
-//   - Valid if it is less than or equal to the last observed block number by this endpoint.
-//   - Invalid if it is greater than the last observed block number by this endpoint.
-func (e *endpoint) validateBlockNumber(perceivedBlockNumber uint64) error {
-	_, err := e.getBlockNumber()
-	if err != nil {
-		return err
-	}
-	// TODO_IMPROVE(@commoddity): implement Allowance for block number check to allow blocks within
-	// a certain range of the current block number to still be considered valid to serve requests.
-	if *e.parsedBlockNumberResponse < perceivedBlockNumber {
-		return fmt.Errorf(errBlockNumberTooLow, e.parsedBlockNumberResponse, perceivedBlockNumber)
+// newEndpoint initializes a new endpoint with the checks that should be run for the endpoint.
+func newEndpoint() endpoint {
+	return endpoint{
+		checkBlockNumber: endpointCheckBlockNumber{},
+		checkChainID:     endpointCheckChainID{},
+		checkArchival:    endpointCheckArchival{},
 	}
-	return nil
 }
 
-// getBlockNumber returns the parsed block number value for the endpoint.
-func (e endpoint) getBlockNumber() (uint64, error) {
-	if e.parsedBlockNumberResponse == nil {
-		return 0, errNoBlockNumberObs
-	}
-	if *e.parsedBlockNumberResponse == 0 {
-		return 0, errInvalidBlockNumberObs
+// getChecks returns the list of checks that should be run for the endpoint on each hydrator run.
+func (e *endpoint) getChecks(es *endpointStore) []gateway.RequestQoSContext {
```

Comment:
Please reconsider: endpoint (and ideally even endpointStore) is best used as a data container only, with serviceState making all the decisions using the data from endpoints (fetched from the endpointStore)

Reply:
Agreed, this logic has been moved to a method on the endpoint store as in your earlier comment.